File: p3/java/app/src/main/java/edu/lehigh/cse262/slang/Slang.java
package edu.lehigh.cse262.slang;

import edu.lehigh.cse262.slang.Env.Env;
import edu.lehigh.cse262.slang.Parser.AstToXml;
import edu.lehigh.cse262.slang.Parser.Parser;
import edu.lehigh.cse262.slang.Scanner.XmlToTokens;

/**
 * JScheme is the entry point into our Scheme parser
 */
public class Slang {
    public static void main(String[] args) {
        var parsedArgs = new Args(args);

        // If help was requested, print help and then exit immediately
        if (parsedArgs.mode == Args.Modes.HELP) {
            Args.printHelp();
            return;
        }

        // Create a SourceLoader and use it to start loading code and scanning /
        // parsing / interpreting it. When given a file, we run exactly one
        // iteration. Otherwise, we run until we get an empty string of text
        // from the sourceLoader.
        //
        // [CSE 262] You should be sure to understand this try-with-resources
        // syntax.
        try (var sourceLoader = new SourceLoader()) {
            // [CSE 262] You should be sure to understand why we make the
            // environment here
            var defaultEnvironment = Env.makeDefault();

            do {
                // Get the code to run, break if no code is available
                String codeToRun;
                if (!parsedArgs.fileName.equals(""))
                    codeToRun = sourceLoader.getFile(parsedArgs.fileName);
                else {
                    System.out.println("Reading input from stdin is not supported in p3.  Please exit and try again");
                    break;
                }
                if (codeToRun.equals(""))
                    break;

                try {
                    // SCAN mode
                    if (parsedArgs.mode == Args.Modes.SCAN) {
                        throw new Exception("-scan is not supported in p3.  Please exit and try again.");
                    }

                    // PARSE mode
                    if (parsedArgs.mode == Args.Modes.PARSE) {
                        var expressions = new Parser(defaultEnvironment.poundT, defaultEnvironment.poundF,
                                defaultEnvironment.empty).parse(XmlToTokens.parse(codeToRun));
                        var xmlPrinter = new AstToXml();
                        for (var expr : expressions)
                            System.out.println(expr.visitString(xmlPrinter));
                    }

                    // INTERPRET mode
                    if (parsedArgs.mode == Args.Modes.INTERPRET) {
                        throw new Exception("-interpret is not supported in p3.  Please exit and try again.");
                    }

                    // FULL mode
                    if (parsedArgs.mode == Args.Modes.FULL) {
                        throw new Exception("-full is not supported in p3.  Please exit and try again.");
                    }
                } catch (Exception e) {
                    // NB: If the scanner, parser, and evaluator are functioning
                    // correctly, the only exceptions that should be thrown are
                    // deliberate, and the proper recovery is to simply print a
                    // message and then resume the repl loop.
                    System.out.println(e.getMessage());
                }
            } while (parsedArgs.fileName.equals("") && false);
        }
    }
}

File: p3/python/slang.py
import sys
import slang_evaluator
import slang_parser
import slang_parser_visitors
import slang_scanner

# This makes exceptions print more nicely.  You should un-comment it when you've
# got your code working:
# sys.tracebacklimit = 0


def getFromStdin(prompt):
    """Read a line from standard input (typically the keyboard).  If nothing is
    provided, return an empty string."""
    try:
        return input(prompt).rstrip()
    except EOFError:
        # we print a newline, because this is going to cause an exit, and it
        # will be a bit more uniform if all ways of exiting have a newline
        # before the shell is back in control
        print("")
        return ""


def getFile(filename):
    """Read a file and return its contents as a single string"""
    source_file = open(filename, "r")
    code = source_file.read()
    source_file.close()
    return code


def printHelp():
    """Print a help message"""
    print("slang.py -- An interpreter for a subset of Scheme, written in Python")
    print("  Usage: python3 slang.py [mode] [filename]")
    print("    * If no filename is given, a REPL will read and evaluate one line of stdin at a time")
    print("    * If a filename is given, the entire file will be loaded and evaluated")
    print("  Modes:")
    print("    -help       Display this message and exit")
    print("    -scan       Scan Scheme code, output tokens as XML")
    print("    -parse      Parse from XML tokens, output an XML AST")
    print("    -interpret  Interpret from XML AST")
    print("    -full       Scan, parse, and interpret Scheme code")


def main(args):
    """Run the Scheme interpreter"""
    # Parse the command-line arguments.  Make sure exactly one valid mode is
    # given, and at most one filename
    filename, mode, numModes, numFiles = "", "", 0, 0
    for a in args:
        if a in ["-help", "-scan", "-parse", "-interpret", "-full"]:
            mode = a
            numModes += 1
        else:
            filename = a
            numFiles += 1
    if numModes != 1 or numFiles > 1 or mode == "-help":
        return printHelp()

    defaultEnv = slang_evaluator.makeDefaultEnv()

    # Run the REPL loop, but only once if we have a valid filename
    while True:
        # Get some code
        codeToRun = ""
        if filename != "":
            codeToRun = getFile(filename)
        else:
            print(
                "Reading input from stdin is not supported in p3.  Please exit and try again")
            break
        if codeToRun == "":
            break

        # Process it according to the mode we're in
        try:
            # SCAN mode
            if mode == "-scan":
                raise Exception("-scan is not supported in p3")

            # PARSE mode
            if mode == "-parse":
                expressions = slang_parser.Parser(defaultEnv.poundT, defaultEnv.poundF, defaultEnv.empty).parse(slang_scanner.XmlToTokens(
                    codeToRun))
                for expr in expressions:
                    print(slang_parser_visitors.AstToXml(expr))

            # INTERPRET mode
            if mode == "-interpret":
                raise Exception("-interpret is not supported in p3")

            # FULL mode
            if mode == "-full":
                raise Exception("-full is not supported in p3")

            # exit if we just processed a file
            if filename != "":
                break

        # any syntax error is going to just print and then exit the interpreter
        except SyntaxError as err:
            print(err.msg)
            break


# In python, this is how we get main() to run when we invoke this program via
# `python3 slang.py ...`
if __name__ == "__main__":
    main(sys.argv[1:])
